compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -226,10 +226,12 @@ installDar depsPath isDataDep ExtractedDar {..} = do
         fp <- dalfFileNameFromEntry dalf
         let targetFp = depsPath </> fp
         let targetDir = takeDirectory targetFp
-        if isDataDep
-          then markDirWith dataDepMarker targetDir
```

Comment:
This line in particular was the culprit. To see why, let's look at the dependency graph for the new test, "data-dependency doesn't leak unexported definitions from transitive dependencies"

```
main          `depends-on` dep
main     `data-depends-on` data-dep
data-dep      `depends-on` dep
```

Even though the dependency from `main` to `dep` is a regular dependency, it was being treated as a data-dependency*. This is because the `.dalf` for `dep` was 
added twice during the flow of `installDependencies`, first from `dep`'s own `.dar` file, and then as one of the included `.dalf`s in `data-dep`'s `.dar` file. Crucially, the second time it was added with the data dependency marker by virtue of being a dependency of a data-dependency. This change ensures that `.dalf`s already in `.daml/dependencies` are not overwritten, so only the markers from the first installation are kept. 

*this can be confirmed by checking that the entry for `dep` under `main/.daml/dependencies/1.14` has a "data dependency marker, i.e. an empty file named `_data_`.